src/wazuh_modules/agent_upgrade/agent/wm_agent_upgrade_com.c

```diff
@@ -0,0 +1,506 @@
+/*
+ * Wazuh Module for Agent Upgrading
+ * Copyright (C) 2015-2020, Wazuh Inc.
+ * July 30, 2020.
+ *
+ * This program is free software; you can redistribute it
+ * and/or modify it under the terms of the GNU General Public
+ * License (version 2) as published by the FSF - Free Software
+ * Foundation.
+ */
+#include <shared.h>
+#include "external/zlib/zlib.h"
+#include "os_crypto/sha1/sha1_op.h"
+#include "os_crypto/signature/signature.h"
+#include "wazuh_modules/wmodules.h"
+#include "wm_agent_upgrade_agent.h"
+
+/**
+ * Static struct to track opened file
+ * */
+static struct {
+    char path[PATH_MAX + 1];
+    FILE * fp;
+} file = {"\0", NULL};
+
+typedef enum _command_error_codes {
+    ERROR_OK = 0,
+    ERROR_UNKNOWN_COMMAND,
+    ERROR_UNSOPPORTED_MODE,
+    ERROR_INVALID_FILE_NAME,
+    ERROR_FILE_OPEN,
+    ERROR_FILE_NOT_OPENED,
+    ERROR_FILE_NOT_OPENED2,
+    ERROR_TARGET_FILE_NOT_MATCH,
+    ERROR_WRITE_FILE,
+    ERROR_CLOSE,
+    ERROR_GEN_SHA1,
+    ERROR_SIGNATURE,
+    ERROR_COMPRESS,
+    ERROR_CLEAN_DIRECTORY,
+    ERROR_UNMERGE,
+    ERROR_CHMOD,
+    ERROR_EXEC,
+    ERROR_CLEAR_UPGRADE_FILE
+} command_error_codes;
+
+static const char * error_messages[] = {
+    [ERROR_OK] = "ok",
+    [ERROR_UNKNOWN_COMMAND] = "Command not found",
+    [ERROR_UNSOPPORTED_MODE] = "Unsupported file mode",
+    [ERROR_INVALID_FILE_NAME] = "Invalid File name",
+    [ERROR_FILE_OPEN] = "File Open Error: %s",
+    [ERROR_FILE_NOT_OPENED] = "File not opened. Agent might have been auto-restarted during upgrade",
+    [ERROR_FILE_NOT_OPENED2] = "No file opened",
+    [ERROR_TARGET_FILE_NOT_MATCH] = "The target file doesn't match the opened file",
+    [ERROR_WRITE_FILE] = "Cannot write file",
+    [ERROR_CLOSE] = "Cannot close file",
+    [ERROR_GEN_SHA1] = "Cannot generate SHA1",
+    [ERROR_SIGNATURE] = "Could not verify signature",
+    [ERROR_COMPRESS] = "Could not uncompress package",
+    [ERROR_CLEAN_DIRECTORY] = "Could not clean up upgrade directory",
+    [ERROR_UNMERGE] = "Error unmerging file",
+    [ERROR_CHMOD] = "Could not chmod",
+    [ERROR_EXEC] = "Error executing command",
+    [ERROR_CLEAR_UPGRADE_FILE] = "Could not erase upgrade_result file"
+};
+
+/**
+ * Format message into the response format
+ * @param error_code code error
+ * @param message string message of the error
+ * @return string meessage with the response format
+ * Response format:
+ * {
+ *  "error": {error_code},
+ *  "data": {message},
+ * }
+ * */
+static char* wm_agent_upgrade_command_ack(int error_code, const char* message);
+
+/**
+ * Process a command that opens a file
+ * @param json_obj expected json format
+ * {
```

Comment:
Done, fut will leave it as 
```
{
       "file":    "file_path",
       "mode":    "wb|w"
 }
```
Since `command `and `parameters `will be parsed in outer function